### ISSUE TITLE
Avoid project loads in find-all-references with plain text searches

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#47656]
When find-all-references is triggered, we will sometimes open up entire projects that reference, or are referenced by, the current project. This means we have to do a full load, parse, and bind of every source-file before we can even traverse the trees to check whether a symbol is applicable.

For example, if you do a find-all-references from `src/compiler`, TypeScript will open up `src/testRunner` to see whether that reference occurred anywhere in the project.

Instead of doing that, when a project is marked as `composite`, we can do a naive text search on each of the files up-front to know if any a candidate occurred in any of them. If none occurred, then we don't have to do an up-front scan/parse/bind, or any checking.

A couple of things to be mindful of:

* There is a risk of doing this over and over again. It's not a great idea to keep hitting disk on every find-all-references if we can, so we shouldn't regress on that.
* If we do have to load the project, it would be nice if we could reuse the candidate list from the initial text search. This is a nice-to-have.

**Comment by amcasey** [comment]
Some editors use caching to make text search faster - maybe we could tap into the editor text search.

**Comment by amcasey** [comment]
Stretch goal: since text search has no cross-file dependencies, it would be cool if we could search more than one at a time (i.e. in parallel).

**Comment by jakebailey** [comment]
Yeah, this is the same optimization that we did in Python on various language servers; go read every file and see if it even contains the text.

**Comment by MichaelMitchell-at** [comment]
One idea would be to let the client, e.g. VSCode, provide the path to a "grep-like" executable which can be used for the file search. Since VSCode comes with ripgrep, it can provide the path to the `ripgrep` binary that it ships with. On MacOS it appears to be at:
`/Applications/Visual Studio Code.app/Contents/Resources/app/node_modules/@vscode/ripgrep/bin/rg`
and Cursor seems to have it at
`/Applications/Cursor.app/Contents/Resources/app/node_modules.asar.unpacked/@vscode/ripgrep/bin/rgArm`
